linker/shared/src/test/scala/org/scalajs/linker/IncrementalTest.scala

```diff
@@ -0,0 +1,142 @@
+        ),
+        classDef(
+            FooClass,
+            kind = ClassKind.ModuleClass,
+            superClass = Some(ObjectClass),
+            memberDefs = List(
+                trivialCtor(FooClass),
+                JSMethodDef(EMF, jsMethodName, Nil,
+                    if (pre) int(5)
+                    else ApplyStatic(EAF, FooClass, staticMethodName, Nil)(IntType)
+                )(EOH, None),
+                MethodDef(EMF.withNamespace(MemberNamespace.PublicStatic),
+                    staticMethodName, NON, Nil, IntType, Some(int(6)))(EOH, None)
+            )
+        )
+    )
+
+    testIncremental(classDefs(_), _ => MainTestModuleInitializers)
+  }
+
+}
+
+object IncrementalTest {
+
+  def testIncremental(
+      classDefs: Boolean => Seq[ClassDef],
+      moduleInitializers: Boolean => List[ModuleInitializer],
+      config: StandardConfig = StandardConfig())(
+      implicit ec: ExecutionContext): Future[Unit] = {
+    Future.sequence(Seq(
+        testIncrementalStep(backward = false, classDefs, moduleInitializers, config),
+        testIncrementalStep(backward = true, classDefs, moduleInitializers, config)
+    )).map(_ => ())
+  }
+
+  private def testIncrementalStep(
+      backward: Boolean,
+      classDefs: Boolean => Seq[ClassDef],
+      moduleInitializers: Boolean => List[ModuleInitializer],
+      config: StandardConfig = StandardConfig())(
+      implicit ec: ExecutionContext): Future[Unit] = {
+
+    val outputInc = MemOutputDirectory()
+    val outputBatch = MemOutputDirectory()
+
+    val linkerInc = StandardImpl.linker(config)
+    val linkerBatch = StandardImpl.linker(config)
+
+    val logger = new ScalaConsoleLogger(Level.Error)
+
+    for {
+      minilib <- TestIRRepo.minilib
+      classDefs0 = minilib ++ classDefs(!backward).map(MemClassDefIRFile(_))
+      classDefs1 = minilib ++ classDefs(backward).map(MemClassDefIRFile(_))
+      inits0 = moduleInitializers(backward)
+      inits1 = moduleInitializers(!backward)
+      _ <- linkerInc.link(classDefs0, inits0, outputInc, logger)
+      reportInc <- linkerInc.link(classDefs1, inits1, outputInc, logger)
+      reportBatch <- linkerBatch.link(classDefs1, inits1, outputBatch, logger)
+    } yield {
+      assertModulesEqual(reportInc.publicModules, reportBatch.publicModules)
+      assertOutputEquals(outputInc, outputBatch)
+    }
+  }
+
+  private def assertModulesEqual(inc: Iterable[Report.Module],
+      batch: Iterable[Report.Module]): Unit = {
+    // Poor man's equality based on toString()
+
+    def strs(ms: Iterable[Report.Module]) =
+      ms.map(m => m.moduleID -> m.toString()).toMap
+
+    assertEquals(strs(inc), strs(batch))
+  }
+
+  private def assertOutputEquals(inc: MemOutputDirectory,
+      batch: MemOutputDirectory): Unit = {
+    val filesInc = inc.fileNames()
+    val filesBatch = batch.fileNames()
+
+    assertEquals(filesInc.toSet, filesBatch.toSet)
+
+    for (f <- filesInc) {
```

Comment:
Consider sorting the list before iterating, so that successive executions of a test reliably show the same error. When trying to fix an issue, if I run a test twice and it gives different errors, I'd like that to be due only to changes I made, and not to non-determinism.
```suggestion
    for (f <- filesInc.sorted) {
```